vaadin-spring/src/main/java/com/vaadin/flow/spring/PrefixTree.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2000-2020 Vaadin Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.vaadin.flow.spring;
+
+import java.io.Serializable;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Quick prefix lookup for package inclusion and exclusion lists.
+ */
+class PrefixTree implements Serializable {
```

Comment:
Random thoughts: Have you estimated the profit in performance of using `PrefixTree` as a replacement to `startsWith`? Just wondering. My concern that the `PrefixTree` solution may bring a memory consumption issues. However it's made only at startup.

Reply:
Profiling on the default (small) list of exclusions indicated only a minor performance improvement of around 15 %, however the previous solution with `startsWith` performed asymptotically poor because it had time complexity `N * M` where `N` is number of classes and `M` the number of packages in the list. With a prefix tree, this is reduced to on average `N * log(M)`. I do no estimate the memory overhead to be too big. A prefix tree may actually save memory in cases where the inclusion / exclusion lists contain many items with a common prefix.

Reply:
Alright, understood. Looks like a good profit! Thanks